### PR TITLE
neuron: cleanup neuron package for upstreaming (WIP)

### PR DESCRIPTION
**Pull request by pramodk**
~- [x] This is still WIP but outlined what "I think" relevant for upstream vs. what is our BBP/developer specific changes.~
~- [x] See my comments as "TODO" and all the code between "=============".~
~- [x] In order to avoid two completely different version in upstream and BBP, I suggest we discuss today the marked sections: @iomaganaris  @matz-e  @ohm314~ 

Here is the updated description:
- [x] NEURON CI for this branch is passing in https://github.com/neuronsimulator/nrn/pull/2636
- [x] @iomaganaris helped to fix the coreneuron perf update CI https://bbpgitlab.epfl.ch/hpc/cellular/coreneuron-perf-tests/-/pipelines/182574
- [x] blueconfig CI is also green now: https://bbpgitlab.epfl.ch/hpc/sim/blueconfigs/-/pipelines/182573

Changes done in this PR:
* remove duplicate neuron recipe in builtin vs patches repo
   * keep BBP dev-specific options in patches
   * the recipe that should go upstream is in the builtin repo  
* remove various legacy logic and variants in the recipe
* just keep the versions that would make sense (especially all alphas)
* some modernisation/refactorings based on the reviews

### CODE REVIEW

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (ohm314) [inline bluebrain/repo-patches/packages/neuron/package.py]: Since the default is reasonable this won't hurt to just stay as an option.

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (ohm314) [inline bluebrain/repo-patches/packages/neuron/package.py]: agree

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (ohm314) [inline bluebrain/repo-patches/packages/neuron/package.py]: i would remove this. I think for the special case where you want to run performance tests you'd want to maybe build on the side, doesn't seem like a thing I'd want in a pkg manager.

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (ohm314) [inline bluebrain/repo-patches/packages/neuron/package.py]: agree

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (matz-e) [inline bluebrain/repo-patches/packages/neuron/package.py]: This is great! I download this and then wildly guess later which software it actually is when I look at `~/Downloads`?

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (matz-e) [inline bluebrain/repo-patches/packages/neuron/package.py]: These seem moot with conditional variants.

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (matz-e) [inline bluebrain/repo-patches/packages/neuron/package.py]: What are these issues? Maybe just remove that comment.

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (matz-e) [inline bluebrain/repo-patches/packages/neuron/package.py]: `define_from_variant`! https://spack.readthedocs.io/en/latest/build_systems/cmakepackage.html#adding-flags-to-cmake

**Review by matz-e — DISMISSED** [review]
This file should not contain the string `"-D`

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (pramodk) [reply bluebrain/repo-patches/packages/neuron/package.py]: I thought the same when I copied the URL 🙄 Agree, we need to fix.

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (pramodk) [reply bluebrain/repo-patches/packages/neuron/package.py]: that's being me not up to date! will do!

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (iomaganaris) [reply bluebrain/repo-patches/packages/neuron/package.py]: I think that this could be helpful to keep in the BBP package only for the ease of use when we have to build neurodamus-packages for reproducing issues, CI, etc

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (iomaganaris) [reply bluebrain/repo-patches/packages/neuron/package.py]: I agree that this shouldn't be in the upstream package. For the BBP CI cases I think this helps a lot to simplify the CIs

**Review by iomaganaris — COMMENTED** [review]
Some comments from quickly looking through the discussion items

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (matz-e) [inline bluebrain/repo-patches/packages/neuron/package.py]: `self.define`...

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (matz-e) [inline bluebrain/repo-patches/packages/neuron/package.py]: `define_from_variant` or similar

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (matz-e) [inline bluebrain/repo-patches/packages/neuron/package.py]: I think this could also be rewritten using `which("awk")(...)`

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (pramodk) [reply bluebrain/repo-patches/packages/neuron/package.py]: @matz-e : could you check if I have updated all `-D..` correctly with `self.define` or equivalent?

File: var/spack/repos/builtin/packages/neuron/package.py
Comment (matz-e) [inline var/spack/repos/builtin/packages/neuron/package.py]: Upstream will tell you this should be version `"master"` last I recall.

File: var/spack/repos/builtin/packages/neuron/package.py
Comment (matz-e) [inline var/spack/repos/builtin/packages/neuron/package.py]: This is all `define_from_variant` :)

File: var/spack/repos/builtin/packages/neuron/package.py
Comment (matz-e) [inline var/spack/repos/builtin/packages/neuron/package.py:169]: I think both file filters here could be merged, too.

File: var/spack/repos/builtin/packages/neuron/package.py
Comment (pramodk) [reply var/spack/repos/builtin/packages/neuron/package.py]: yeah, I remember and it's a bit annoying to have that changed. I guess there will be some more changes from the upstream review. After that, I pull the upstream recipe into our fork.

File: var/spack/repos/builtin/packages/neuron/package.py
Comment (pramodk) [reply var/spack/repos/builtin/packages/neuron/package.py:169]: do you meant something like:

```console
-        if spec.satisfies("@:7"):
-            assign_operator = "?="
-        else:
-            assign_operator = "="
-
-        filter_file(
-            "CC {0} {1}".format(assign_operator, env["CC"]),
-            "CC = {0}".format(cc_compiler),
-            nrnmech_makefile,
-            **kwargs,
-        )
-
-        filter_file(
-            "CXX {0} {1}".format(assign_operator, env["CXX"]),
-            "CXX = {0}".format(cxx_compiler),
-            nrnmech_makefile,
-            **kwargs,
-        )
+        assign_operator = "?=" if spec.satisfies("@:7") else "="
+
+        # older neuron versions also use CC
+        compilers = [("CC", "cc_compiler"), ("CXX", "cxx_compiler")]
+
+        for compiler_var, compiler_env in compilers:
+            pattern = "{0} {1} {2}".format(
+                compiler_var, assign_operator,  env[compiler_env]
+            )
+            replacement = "{0} = {1}".format(
+                compiler_var, locals()[compiler_env]
+            )
+            filter_file(pattern, replacement, nrnmech_makefile, **kwargs)
```

I am also trying to not add too many changes/improvements to avoid breaking something (this makefile replacement on cray vs non-cray platforms was buggy)

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (pramodk) [inline bluebrain/repo-patches/packages/neuron/package.py]: ⚠️ : I will replace this commit after neuronsimulator/nrn/pull/2636 is merged!

**Review by iomaganaris — DISMISSED** [review]
As far as I can see LGTM 👍 
Very welcome improvement!

**Comment by pramodk** [comment]
The above pipeline failed because I manually canceled the running job so that we could launch a fresh build one more time before merging.

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (heerener) [inline bluebrain/repo-patches/packages/neuron/package.py]: `s/furing/during/`
But please do this in a future PR, don't hold back this PR for a nitpicky non-issue!

**Review by heerener — DISMISSED** [review]
Looks a lot more digestible than before!

Please also make sure to get an approval from the spack-master himself :-)

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (pramodk) [reply bluebrain/repo-patches/packages/neuron/package.py]: thanks for spotting that! 😒 

I will see if I need to push another commit otherwise I will merge it.

A bit painful thing has been that neuron-check CI is failing "randomly" only with GPU builds. If I retrigger that on different  GPU nodes (especially non-ldu* / phase2) then everything is passing).

I am checking this by explicitly launching neuron CI on this branch: https://github.com/neuronsimulator/nrn/pull/2646.

If that will be green and if re-triggered CI is green than I will merge this.

Comment (pramodk) [reply bluebrain/repo-patches/packages/neuron/package.py]: addressed as I need to fix something anyway.

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (matz-e) [inline bluebrain/repo-patches/packages/neuron/package.py]:
```suggestion
        args = super().cmake_args()
```
Python 3 only by now.

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (matz-e) [inline bluebrain/repo-patches/packages/neuron/package.py:132]: We can't handle floating point handling in CMake?

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (matz-e) [inline bluebrain/repo-patches/packages/neuron/package.py]: Still a `-D` :)

File: bluebrain/spack-scripting/scripting/cmd/configure_pipeline.py
Comment (matz-e) [inline bluebrain/spack-scripting/scripting/cmd/configure_pipeline.py:1]: Do we need the changes here, still? Could we not define the bogus `develop` version only in our package and then skip building upon the card house created by this file?

File: var/spack/repos/builtin/packages/neuron/package.py
Comment (matz-e) [reply var/spack/repos/builtin/packages/neuron/package.py]: I'd move this to our `patch` repo.

File: var/spack/repos/builtin/packages/neuron/package.py
Comment (matz-e) [reply var/spack/repos/builtin/packages/neuron/package.py:169]: Yes. as an added bonus, in the new deployment, no lines match `env[compiler_env]`, and I just matched `^CXX {assign_operator} .*` to get things working "better". Still broken, though. Upstream `neuron` will be pained by the same.

**Review by matz-e — DISMISSED** [review]

File: var/spack/repos/builtin/packages/neuron/package.py
Comment (pramodk) [reply var/spack/repos/builtin/packages/neuron/package.py]: done!

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (pramodk) [reply bluebrain/repo-patches/packages/neuron/package.py]: done in ff51c40c6b187e9c98e804b1bfa3c3a58e1af6a8

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (pramodk) [reply bluebrain/repo-patches/packages/neuron/package.py]: coffee-pasta 😒. Done!

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (pramodk) [reply bluebrain/repo-patches/packages/neuron/package.py:132]: note that this is for `@:9.0.a14` and the CMake change to handle such flags was added only recently. In neurodamus we often need to install older neuron versions for result cross-checking and hence I thought it would be helpful to have this.

File: bluebrain/spack-scripting/scripting/cmd/configure_pipeline.py
Comment (pramodk) [reply bluebrain/spack-scripting/scripting/cmd/configure_pipeline.py:1]: A valid point seeing this diff! I have no background of this `cmd/configure_pipeline.py` and wondering how you allowed to build this `card house` in the first place 😄 

@iomaganaris added this to fix coreneuron perf CI. I have to differ to him as I also don't know all the aspect about the perf-measurement CI.

File: var/spack/repos/builtin/packages/neuron/package.py
Comment (pramodk) [reply var/spack/repos/builtin/packages/neuron/package.py:169]: Done

**Review by pramodk — COMMENTED** [review]
Addressed last comments from @matz-e. I will defer the aspect about `/cmd/configure_pipeline.py` as I don't have sufficient background. Otherwise everything should be in place!

File: bluebrain/spack-scripting/scripting/cmd/configure_pipeline.py
Comment (iomaganaris) [reply bluebrain/spack-scripting/scripting/cmd/configure_pipeline.py:1]: The problem I tried to address here was an issue with the `configure-pipeline` command due to the fact that `version("develop")` is not defined in the patch recipe of NEURON since it's defined in the `builtin` recipe . I thought I would make more sense to add a `develop` version automatically if this command is used since there might be more recipes written without a `develop` version defined for some reason (either because it's a recipe built on top of a `builtin` one or there is `version("master"` instead). If you think it's not a good idea we could also define a `develop` version in the `patch` recipe of NEURON but there is a small chance we come across this issue again.

Edit: I saw you moved the `version("develop")` to the patch recipe so this fix shouldn't be needed anymore however I would propose throwing an error in case `develop` is not found in the recipe so we don't just continue building stuff with the incorrect branches, etc. (Could be done in a separate PR as well)

Comment (pramodk) [reply bluebrain/spack-scripting/scripting/cmd/configure_pipeline.py:1]: thanks for the explanation, Ioannis! (By tagging, didn't mean to bother you! we revisit this after the break!)

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (matz-e) [inline bluebrain/repo-patches/packages/neuron/package.py:104]: This can't be simplified?

File: bluebrain/spack-scripting/scripting/cmd/configure_pipeline.py
Comment (matz-e) [inline bluebrain/spack-scripting/scripting/cmd/configure_pipeline.py:1]: I hope these modifications are not needed any longer?

**Review by matz-e — COMMENTED** [review]
Looks good, last open point for me is that the `configure_pipeline` should really not need modifications (we should replace that if at all)

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (pramodk) [reply bluebrain/repo-patches/packages/neuron/package.py:104]: Do you mean avoiding multi-valued variant altogether? that's possible but then we will remove flexibility and compile all the tests every time (which adds build time).

I would suggest to keep as it is unless @iomaganaris has different idea.

File: bluebrain/spack-scripting/scripting/cmd/configure_pipeline.py
Comment (matz-e) [reply bluebrain/spack-scripting/scripting/cmd/configure_pipeline.py:1]: I just checked, and our new spack has a fix for installing `neuron@git.<sha>=develop`. This should make parsing recipes and modifying them a moot point once we merge the next deployment iteration.

Until then, no further complications to this script, please :)

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (iomaganaris) [inline bluebrain/repo-patches/packages/neuron/package.py:1]: Nitpick
```suggestion
# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
```

File: var/spack/repos/builtin/packages/neuron/package.py
Comment (iomaganaris) [inline var/spack/repos/builtin/packages/neuron/package.py:1]: Nitpick
```suggestion
# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
```

**Review by iomaganaris — APPROVED** [review]

File: bluebrain/repo-patches/packages/neuron/package.py
Comment (pramodk) [reply bluebrain/repo-patches/packages/neuron/package.py:1]: @iomaganaris: 2023 -> 2024 changes would be part of upstreaming merge I believe. So I will keep it as it is.

**Review by matz-e — APPROVED** [review]

**Pull request merged**
